src/app/components/autocomplete/autocomplete.component.ts

```diff
@@ -0,0 +1,156 @@
+	providers: [
+		{
+			provide: NG_VALUE_ACCESSOR,
+			useExisting: forwardRef(() => AutocompleteComponent),
+			multi: true,
+		},
+		{
+			provide: NG_VALIDATORS,
+			useExisting: forwardRef(() => AutocompleteComponent),
+			multi: true,
+		},
+	],
+})
+export class AutocompleteComponent implements ControlValueAccessor, Validators {
+	/*
+    keep track of package names that has been already loaded and show them
+    as suggestion for the user
+  */
+	@Input() autocompletePackageNames: string[] | null = [];
+
+	/*
+    Pacakge names that are already loaded to prevent selecting same library
+    multiple times
+  */
+	private alreadyLoadedPackageName: ArrayObservable<string> = new ArrayObservable();
+
+	/* Observale that will display loaded packages from NPM */
+	readonly autocompleteOptions$!: Observable<string[]>;
+
+	/* Form control to allow user search packages  */
+	readonly addPackage: FormControl<string> = new FormControl('', {
+		asyncValidators: this.errorHandlerService.noDuplicatesValidator(this.alreadyLoadedPackageName.observable$),
+		nonNullable: true,
+	});
+
+	/* errors that may occur when searching for a package name */
+	readonly packageErrorsHandler = this.errorHandlerService.getInputErrorsHandler('package name');
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	onChange = (value: string[]) => {
+		/* empty */
+	};
+	onTouched = () => {
+		/* empty */
+	};
+	constructor(
+		private readonly npmRegistryService: NpmRegistryService,
+		private readonly errorHandlerService: ErrorHandlerService
+	) {
+		// npm library suggestions on user input
+		const suggestions$ = this.searchLibraryOnInputChange();
+
+		this.autocompleteOptions$ = combineLatest([
+			// All suggestions
+			suggestions$.pipe(startWith([])),
+			// packages that already have been loaded
+			this.alreadyLoadedPackageName.observable$,
+			// Partial npm package name to filter options
+			this.addPackage.valueChanges.pipe(startWith('')),
+		]).pipe(
+			map(([suggestions, alreadyLoadedPackageName, query]) =>
+				this.getAutocompleteOptions(
+					[...new Set([...(this.autocompletePackageNames ?? []), ...suggestions])],
+					alreadyLoadedPackageName,
+					query
+				)
+			)
+		);
+	}
+
+	writeValue(value?: string[]): void {
+		this.alreadyLoadedPackageName.set([...(value ?? [])]);
+	}
+	registerOnChange(fn: (value: string[]) => void): void {
+		this.onChange = fn;
+	}
+	registerOnTouched(fn: (value: void) => void): void {
+		this.onTouched = fn;
+	}
+
+	validate(c: AbstractControl): ValidationErrors | null {
+		if (!this.addPackage.validator) {
+			return null;
+		}
+
+		return this.addPackage.validator(c);
+	}
+
+	onSubmit(): void {
+		this.alreadyLoadedPackageName.push(this.addPackage.value);
+		this.onChange(this.alreadyLoadedPackageName.getValue());
+
+		// Clear value
+		this.addPackage.setValue('');
+		this.addPackage.markAsUntouched();
+	}
+
+	private searchLibraryOnInputChange(): Observable<string[]> {
```

Comment:
`searchLibraryOnInputChange()` and `getAutocompleteOptions()` should stay in `AppComponent` as well